Another series of cmake simplifications

arnout opened the pull request:
The idea is to remove multiap-helpers.cmake and define everything in top-level CMakeLists.txt.

Also to use standard PROJECT_VERSION etc. instead of custom things like FOO_VERSION_STRING.

arnout committed:
cmake: set (top-level) VERSION in project()

To be able to do this, policy CMP0048 must be set to NEW. For
simplicity, we set everything introduced in CMake 3.0 to NEW.

Note that this triggers CMake warnings on all subdirectories that use
project() without explicit VERSION, because they override the
PROJECT_VERSION variables to an empty string. This will be fixed in a
later commit.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: simplify some common options

Instead of setting PROJECT_NAME, just use project() (it sets
PROJECT_NAME).

Remove the cmake_minimum_version, it's already done at top level.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on CMakeLists.txt:
```diff
@@ -1,5 +1,8 @@
+cmake_policy(VERSION 3.0)
+project (prplmesh
+         LANGUAGES C CXX
+         VERSION 1.4.0)
```
I think we should set the version to 1.0 now. We refer to it as `1.0` in other places, so its confusing.

arnout replied:
> I think we should set the version to 1.0 now. We refer to it as `1.0` in other places, so its confusing.

I agree, however I prefer to do that in a separate commit/PR (first make sure that no references to 1.4.0 remain).

Also, should we move to 1.0.0, 1.0, 1.99.0, or 2.0.0?

tomereli commented on CMakeLists.txt:
```diff
@@ -4,7 +4,51 @@ project (prplmesh
          VERSION 1.4.0)
 
-include(cmake/multiap-helpers.cmake)
+if(NOT WIN32)
```
This actually beats the purpose of multiap-helpers.cmake which aims to keep the `CMakeLists.txt` cleaner, can't say I like this. What's the harm with redefining these variables? If there is a real reason to not do this, I think we can add something like header guards for cmake includes, can't we?

vitalybu replied:
Why not simply include once from the top cmake?
Not sure there's a reason to include it multiple times after we switched to a single cmake project.

arnout replied:
I'll remove the first commit. The main purpose was to avoid including it multiple times (with different PROJECT values), but it's true that this was not the best approach.

tomereli commented on agent/src/beerocks/monitor/CMakeLists.txt:
```diff
@@ -1,8 +1,4 @@
-
-set(PROJECT_NAME beerocks_monitor)
-project (${PROJECT_NAME})
-message("${BoldWhite}Preparing ${BoldGreen}${PROJECT_NAME}${BoldWhite} for the ${BoldGreen}${TARGET_PLATFORM}${BoldWhite} platform${ColourReset}")
```
I disagree that it's just noise - it helps developers spot the areas in the configuring stage. I prefer to keep it, but I'd like to hear what @vitalybu thinks.

vitalybu replied:
I agree with @tomereli.
prplMesh contains several components and it does help to see the progress of the configuration stage.

arnout replied:
OK I'll keep the message calls.

It was more of an accident - I had removed it in 3 files and then I thought I can just as well continue doing it :-)

arnout committed:
cmake: use BUILD_DATE instead of ${PROJECT}_BUILD_DATE

Since we don't have 4 separate projects any more, it doesn't make sense
to compute the build date 4 times. Therefore, just compute it once.

Note that since multiap-helpers.cmake is still included 4 times, there
will in fact be 4 different build dates. A later commit will remove
that.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: allow BUILD_DATE to be set externally

For reproducible builds, we'll want to set BUILD_DATE externally. Allow
this by only setting it if not already set.

This has the side effect of not having 4 different build dates, which is
exactly what we want.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: use REVISION instead of ${PROJECT}_REVISION

Since we don't have 4 separate projects any more, it doesn't make sense
to compute the git revision 4 times. Therefore, just compute it once, by
using a REVISION variable without prefix.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: include multiap-helpers.cmake only once

For historical reasons, multap-helpers.cmake was included from each
component, and again from the top-level CMake. This is pretty redundant.
When we set PROJECT to a single value (PRPLMESH) from the top level, we
can use it everywhere.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: use standard prplmesh_VERSION_MAJOR

The project(prplmesh VERSION 1.4.0) command automatically creates a
variable prplmesh_VERSION_MAJOR. Use that instead of coding it
explicitly in multiap-helpers.cmake.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: use standard prplmesh_VERSION instead of _VERSION_STRING

The project(prplmesh VERSION 1.4.0) command automatically creates a
variable prplmesh_VERSION. Use that instead of coding it explicitly in
multiap-helpers.cmake.

Note that PRPLMESH_VERSION_MAJOR is still used in some places and will
be replaced later, so that is still defined explicitly.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: remove PROJECT variable

It is no longer used for anything significant.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: move setting of output directory to top level

The four component CMakeLists.txt set this in exactly the same way, so
just move it to top level.

Also move the message() calls to top-level, to reduce the noise.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: move setting of C,CXX,LINK flags to top-level

The four components repeat the setting of CFLAGS, CXXFLAGS and
EXE_LINKER_FLAGS in slightly different ways. Move all of this to the
top-level CMakeLists.txt and unify them.

This entails the following changes.

agent:
- set both CFLAGS and CXXFLAGS;
- -std=c++11 is set with CMAKE_CXX_STANDARD, if supported;
- -s flag is passed through _RELEASE flags instead of through a
condition;
- -Wno-comment is added;
- -Wno-address-of-packed-member only passed for GCC 4.9+.

common:
- -std=c++11 is set with CMAKE_CXX_STANDARD, if supported;
- -fPIC not passed for non-RELEASE (it actually still is passed because
CMake adds it automatically when building a shared library);
- -fPIC is passed explicitly in _RELEASE flags - we want both PIC and
PIE and for static libs PIC isn't added automatically;
- Remove some redundant messages (already reported at top-level);
- also add rpath manipulation to linker flags.

controller:
- -std=c++11 is set with CMAKE_CXX_STANDARD, if supported;
- -fPIC not passed for non-RELEASE (it actually still is passed because
CMake adds it automatically when building a shared library);
- -fPIC is passed explicitly in _RELEASE flags - we want both PIC and
PIE and for static libs PIC isn't added automatically.

framework:
- remove -fpic, redundant with -fPIC;
- don't force CMAKE_CXX_FLAGS and don't put it in cache (it isn't by
default);
- Remove some redundant messages (already reported at top-level);
- -Wno-address-of-packed-member only passed for GCC 4.9+.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: cmake: remove unused BWL/BPL_MODULE_NAME

Since commit b169ac1180, BPL_MODULE_NAME is no longer used and since
commit d207867fa10e4, BWL_MODULE_NAME is no longer used.

Remove their definition in CMakeLists.txt, and the FIXME comment which
is no longer relevant.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: set minimum CMake to 3.1

This allows us to use compile features.

For now, just set CXX standard to 11. Also use
CMAKE_CXX_STANDARD_REQUIRED to check the compiler version instead of an
explicit version check.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
framework: cmake: remove -std=c++11

It is already passed at top-level, no need to repeat it.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: remove -fpermissive

-fpermissive lowers the impact of some errors to warning. Since we build
with -Werror, that is kind of useless. Remove it.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: cmake: move -Wno-shift-negative-value down to cli

The -Wno-shift-negative-value warning suppression is only needed in one
specific case: with some toolchains (on ARM), the warning is issued due
to the static cast from derived class member function to parent class
member function. We can be sure that the warning is harmless in that
particular case, but not really in general.

Therefore, move the warning down to the specific compile that needs it.

Also use add_compile_options() instead of messing with CMAKE_C_FLAGS.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on CMakeLists.txt:
```diff
@@ -1,5 +1,8 @@
 cmake_minimum_required(VERSION 3.0)
-project (prplmesh C CXX)
+cmake_policy(VERSION 3.0)
```
Perhaps it would be better to add a comment here with the link to https://cmake.org/cmake/help/v3.0/policy/CMP0048.html
(nitpick since it is part of the commit message)

arnout replied:
`cmake_policy(VERSION ...)` is a shorthand to avoid requiring an explicit setting for each policy. It is advised in cmake docs to always set it.

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -40,7 +40,7 @@ endif()
 
 include(multiap-helpers)
 add_definitions(-DBEEROCKS_VERSION="${${PROJECT}_VERSION_STRING}")
```
So shouldn't we do the same for the version string and revision? PROJECT is always `prplMesh`.

tomereli commented on cmake/multiap-helpers.cmake:
```diff
@@ -41,7 +41,11 @@ set(${PROJECT}_VERSION_MAJOR 1)
 set(${PROJECT}_VERSION_STRING ${${PROJECT}_VERSION_MAJOR}.${${PROJECT}_VERSION_MINOR}.${${PROJECT}_VERSION_PATCH})
-execute_process(COMMAND "date" "+%F %T" OUTPUT_STRIP_TRAILING_WHITESPACE OUTPUT_VARIABLE BUILD_DATE)
+
+if (NOT BUILD_DATE)
```
But if we just include this file once from the top level CMakeLists.txt, we wouldn't need to check it, would we?

arnout replied:
No, the idea of the condition is to be able to override it from the command line. The problem with a build date is that when you build twice in a row, you get two different binaries because the build date is different.

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -41,7 +41,7 @@ endif()
 include(multiap-helpers)
 add_definitions(-DBEEROCKS_VERSION="${${PROJECT}_VERSION_STRING}")
 add_definitions(-DBEEROCKS_BUILD_DATE="${BUILD_DATE}")
-add_definitions(-DBEEROCKS_REVISION="${${PROJECT}_REVISION}")
```
I think a preparative commit which moves all these definitions to the top level CMakeLists.txt makes more sense. Then we can move it to the multiap_helpers.cmake and be done with it.

arnout replied:
Meh, that's a lot of rebase work :-)

I wasn't sure about moving it up because it's only used in two components.

tomereli replied:
should have been a (nitpick) anyway, don't worry about it

tomereli commented on cmake/multiap-helpers.cmake:
```diff
@@ -29,12 +29,12 @@ if (${FORCE_COLORED_OUTPUT})
 endif ()
 
-if (NOT ${PROJECT}_REVISION)
+if (NOT REVISION)
```
PRPLMESH_REVISION is more appropriate IMO (nitpick)

arnout replied:
Yes I considered that. But then I didn't think the prefix is very useful - we're in the prplmesh project, so obviously it's about prplmesh... Why don't we have PRPLMESH_INSTALL_PATH then for example?

tomereli commented on agent/config/beerocks_agent_version.in:
```diff
@@ -1,2 +1,2 @@
-@PRPLMESH_VERSION_STRING@
+@prplmesh_VERSION@
 @REVISION@
```
Seeing this, I guess `prplmesh_REVISION` is more appropriate.

tomereli reviewed (approved):
I think that commit 6f834537b45dcf58df8bdd388e45ff29196198d6 should be a preparative commit, then we won't need the workarounds done in the commit c01382fc6d4c64849c23c1854f767bf50f0ef9af.

Other than that, and provided that this doesn't break UGW build, I don't have anything critical so approving already.
